### PR TITLE
SONARJAVA-3749 Rule S6205: Switch arrow labels should not use redundant keywords

### PR DESCRIPTION
**Pull request by quentin-jaquier-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    StatementTree lastStatement = body.get(statementsInBody - 1);
+    boolean isLastStatementBreak = lastStatement.is(Tree.Kind.BREAK_STATEMENT);
+
+    if (statementsInBody == 1) {
+      if (lastStatement.is(Tree.Kind.YIELD_STATEMENT)) {
+        reportIssue(blockTree, String.format(MESSAGE, "block and \"yield\""));
+      } else {
+        reportIssue(blockTree.openBraceToken(), String.format(MESSAGE, "block"));
+      }
+    } else if (isLastStatementBreak) {
```
Comment (margarita-nedzelska-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java:67]: What about checking  `lastStatement.is(Tree.Kind.BREAK_STATEMENT)` right in the `if` clause?

File: java-checks/src/test/java/org/sonar/java/checks/SwitchRedundantKeywordCheckTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.junit.jupiter.api.Test;
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+import static org.sonar.java.checks.verifier.TestUtils.nonCompilingTestSourcesPath;
+
+class SwitchRedundantKeywordCheckTest {
```
Comment (margarita-nedzelska-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/SwitchRedundantKeywordCheckTest.java:27]: Maybe I would suggest testing on Java 8 version too.

**Review by margarita-nedzelska-sonarsource — COMMENTED** [review]
Overall looks good, just few suggestions

File: java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  private void reportRedundantKeywords(BlockTree blockTree) {
+    List<StatementTree> body = blockTree.body();
+    int statementsInBody = body.size();
+    StatementTree lastStatement = body.get(statementsInBody - 1);
+
+    if (statementsInBody == 1) {
+      if (lastStatement.is(Tree.Kind.YIELD_STATEMENT)) {
+        reportIssue(blockTree, String.format(MESSAGE, "block and \"yield\""));
```
[the remark this answers is not in the document]
Comment (quentin-jaquier-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java]: Make sense, I went for reporting the redundant keyword, and the braces as secondary locations.

File: java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      }
+    }
+  }
+
+  private void reportStatementInBlock(Tree statement, BlockTree blockTree, String redundantParts) {
+    reportIssue(statement,
+      String.format(MESSAGE, redundantParts),
+      Arrays.asList(new JavaFileScannerContext.Location("Redundant opening brace", blockTree.openBraceToken()),
+        new JavaFileScannerContext.Location("Redundant closing brace", blockTree.closeBraceToken())),
+      null);
```
Comment (margarita-nedzelska-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java:93]: Is it correct indentation?

File: java-checks/src/test/java/org/sonar/java/checks/SwitchRedundantKeywordCheckTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  @Test
+  void test_java_version_8() {
+    JavaCheckVerifier.newVerifier()
+      .onFile(testSourcesPath("checks/SwitchRedundantKeywordCheck.java"))
+      .withJavaVersion(8)
+      .withCheck(new SwitchRedundantKeywordCheck())
+      .verifyNoIssues();
+  }
+
+
```
Comment (margarita-nedzelska-sonarsource) [inline java-checks/src/test/java/org/sonar/java/checks/SwitchRedundantKeywordCheckTest.java:58]: Looks like too many empty lines at the end :)

**Review by margarita-nedzelska-sonarsource — APPROVED** [review]
Looks good to me. Few style comments.
Nice work!

File: java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java (hunk 4 above)
Comment (quentin-jaquier-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/SwitchRedundantKeywordCheck.java]: I'm afraid it is, at least what is suggested by IntelliJ. In any case, I extracted this code into a separate method to make the whole cleaner.

**Pull request merged**